src/ui/public/courier/data_source/_abstract.js

```diff
@@ -287,20 +287,26 @@ export function AbstractDataSourceProvider(Private, Promise, PromiseEmitter) {
           };
         }
 
-        if (flatState.body.size > 0) {
-          const computedFields = flatState.index.getComputedFields();
-          flatState.body.stored_fields = computedFields.storedFields;
-          flatState.body.script_fields = flatState.body.script_fields || {};
-          flatState.body.docvalue_fields = flatState.body.docvalue_fields || [];
-
-          _.extend(flatState.body.script_fields, computedFields.scriptFields);
-          flatState.body.docvalue_fields = _.union(flatState.body.docvalue_fields, computedFields.docvalueFields);
-
-          if (flatState.body._source) {
-            // exclude source fields for this index pattern specified by the user
-            const filter = fieldWildcardFilter(flatState.body._source.excludes);
-            flatState.body.docvalue_fields = flatState.body.docvalue_fields.filter(filter);
-          }
+        const computedFields = flatState.index.getComputedFields();
+        flatState.body.stored_fields = computedFields.storedFields;
+        flatState.body.script_fields = flatState.body.script_fields || {};
+        flatState.body.docvalue_fields = flatState.body.docvalue_fields || [];
+
+        _.extend(flatState.body.script_fields, computedFields.scriptFields);
+        flatState.body.docvalue_fields = _.union(flatState.body.docvalue_fields, computedFields.docvalueFields);
+
+        if (flatState.body._source) {
+          // exclude source fields for this index pattern specified by the user
+          const filter = fieldWildcardFilter(flatState.body._source.excludes);
+          flatState.body.docvalue_fields = flatState.body.docvalue_fields.filter(filter);
+        }
+
+        // if we only want to search for certain fields
+        const fields = flatState.fields;
+        if (fields) {
+          flatState.body.docvalue_fields = _.intersection(flatState.body.docvalue_fields, fields);
+          flatState.body.script_fields = _.pick(flatState.body.script_fields, fields);
+          flatState.body._source = _.difference(fields, [...flatState.body.docvalue_fields, ..._.keys(flatState.body.script_fields)]);
```

Comment:
> This is currently being hard-coded to ['*'] here. I could potentially include the same fields that we're requesting from _source here, do you think that'd help performance?

It could, if someone has some very large stored fields that don't need to be pulled. I don't think it could hurt, could it? If you include a field that is not actually configured to be `stored` in the list, Elasticsearch will simply ignore it.